src/serverlist.cpp

```diff
@@ -649,6 +649,14 @@ void CServerListManager::RetrieveAll ( const CHostAddress& InetAddr )
         // if the client IP address is a private one, it's on the same LAN as the directory
         bool clientIsInternal = NetworkUtil::IsPrivateNetworkIP ( InetAddr.InetAddr );
 
+        CHostAddress clientPublicAddr = CHostAddress ( InetAddr.InetAddr, InetAddr.iPort );
+        if ( clientIsInternal && CHostAddress().InetAddr != ServerList[0].LHostAddr.InetAddr &&
+             !NetworkUtil::IsPrivateNetworkIP ( ServerList[0].LHostAddr.InetAddr ) )
+        {
+            // client and directory on same LAN, directory has public IP set, use that for client public IP
+            clientPublicAddr.InetAddr = ServerList[0].LHostAddr.InetAddr;
```

Comment:
Is this always safe? Could we get the public address from the client directly?

Reply:
The point here is that the client doesn't know its public IP address - it's on the same LAN as the directory.  The client _never_ knows its public IP address.

And hence, where the `if` test is true, yes, this is the best we can manage: the client is on the same LAN, the LHostAddr isn't "default" (unusable) and it's also not on the same LAN (unuseable).  It's set in the class constructor, either from `--serverpublicip` or from autodetection (which will get a value `IsPrivateNetworkIP` says "yes" to, unless it's actually useable).

At the worst case, we end up using a value that's no use and the client can't connect -- i.e. exactly the same as it is now.

Reply:
Ok. Maybe the comment should then say: ... use that for client public IP as it is most likely the same as the client.

Reply:
Done.

Reply:
Thanks.